foundationdb/src/tuple.rs

```diff
@@ -8,20 +8,81 @@
 
 //! Tuple Key type like that of other FoundationDB libraries
 
-use byteorder::{self, ByteOrder};
 use std;
 use std::io::Write;
 
-#[derive(Debug, Clone, Copy)]
+use byteorder::{self, ByteOrder};
+
+#[derive(Debug, Fail)]
 pub enum TupleError {
-    //
+    #[fail(display = "Unexpected end of file")]
     EOF,
-    InvalidType,
+    #[fail(display = "Invalid type")]
+    InvalidType(u8),
+    #[fail(display = "Invalid data")]
     InvalidData,
 }
 
 type Result<T> = std::result::Result<T, TupleError>;
 
+/// Various tuple types
+#[derive(Clone, Copy, Eq, PartialEq)]
+enum SingleType {
```

Comment:
It seems that there's little benefit for using enum for this case. Using enumuration compicates things by enforcing `as u8` cast everywhere, while `try_from` is only used on `Single::decode`. How about using `const TYPE_EMPTY: u8 = 0x00`?

Reply:
> How about using `#[repr(u8)]`?

Fantastic recommendation!

> It seems that there's little benefit for using enum for this case

The benefits I see are this:
- each SingleType code is unique and should not overlap, enum captures that
- in match clauses we get uncovered case errors during compilation.
- It allows us to associate custom functionality, which we'd have to do with a Trait + impl for u8, this is a little cleaner than that.

If you think this generally makes the code harder to follow, and doesn't do much to reduce errors, then I'm happy to switch to named const's, personally I like using this patter in cases like this, but I'm not going to push it.

Reply:
I think we'd better to go with named constant.

The second one is the most beneficial one as it catches tons of human errors. However in this case, we need to use wildcard match anyway because range of header values from `NegIntStart` to `PosIntEnd` maps to integer type.
The first one could be covered by compiler warnings on match clause. For example,
https://play.rust-lang.org/?gist=8769a3fcd215ba9a08b0b3ec32e5654d&version=stable&mode=debug

Reply:
Ok, I've converted to named consts.